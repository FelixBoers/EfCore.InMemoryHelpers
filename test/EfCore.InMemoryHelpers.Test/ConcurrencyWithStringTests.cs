using ApprovalTests;
using Microsoft.EntityFrameworkCore;
using Xunit;
using Xunit.Abstractions;

namespace EfCore.InMemoryHelpers.Test
{
    public class ConcurrencyWithStringTests : TestBase
    {
        public ConcurrencyWithStringTests(ITestOutputHelper output)
            :
            base(output)
        { }

        [Fact]
        public void NullConflictThrows()
        {
            var entity = new TestEntity
            {
                Property = "prop",
                ConcurrencyToken = "573dfb98237e4f51bafaab0317777960"
            };

            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                context.Add(entity);
                context.SaveChanges();
            }

            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                var update = new TestEntity
                {
                    Id = entity.Id,
                    Property = "Something new",
                    ConcurrencyToken = "4b39590ea0894ab9a05912456bf700b6"
                };
                context.Attach(update);
                context.Entry(update).Property("Property").IsModified = true;
                var exception = Assert.Throws<DbUpdateConcurrencyException>(() => context.SaveChanges());
                Approvals.Verify(exception.Message);
            }
        }

        [Fact]
        public void SetConcurrencyTokenToNullThrows()
        {
            var entity = new TestEntity
            {
                Property = "prop",
                ConcurrencyToken = "616e2ca1a52a4c57afbfe80ecc54125c"
            };

            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                context.Add(entity);
                context.SaveChanges();
            }

            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                var update = new TestEntity
                {
                    Id = entity.Id,
                    Property = "Something new",
                    ConcurrencyToken = null
                };
                context.Attach(update);
                context.Entry(update).Property("Property").IsModified = true;
                var exception = Assert.Throws<DbUpdateConcurrencyException>(() => context.SaveChanges());
                Approvals.Verify(exception.Message);
            }
        }

        [Fact]
        public void UpdateMultipleSameEntity()
        {
            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                var entity1 = new TestEntity
                {
                    Property = "prop",
                    ConcurrencyToken = "3c615ac3734046bc8bf6fd58d979c1ae"
                };
                var entity2 = new TestEntity
                {
                    Property = "prop",
                    ConcurrencyToken = "cc4c1ce921e6457b89d00c7c2c6de2a9"
                };
                context.AddRange(entity1, entity2);
                context.SaveChanges();
                entity1.Property = "Something new";
                entity2.Property = "Something new";
                context.SaveChanges();
            }
        }

        [Fact]
        public void UpdateSameEntity()
        {
            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                var entity = new TestEntity
                {
                    Property = "prop",
                    ConcurrencyToken = "28f4fc79da8f43b887d2eef4e82a2be5"
                };
                context.Add(entity);
                context.SaveChanges();
                entity.Property = "Something new";
                context.SaveChanges();
            }
        }

        [Fact]
        public void UpdateSucceeds()
        {
            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                var entity = new TestEntity
                {
                    Property = "prop",
                    ConcurrencyToken = "d6f321a1320846a49b81595c4d41d25d"
                };
                context.Add(entity);
                context.SaveChanges();
                var firstToken = entity.ConcurrencyToken.ToString();
                entity.Property = "Something new";
                context.SaveChanges();
                Assert.Equal(firstToken, entity.ConcurrencyToken);
            }
        }

        [Fact]
        public void ValueConflictThrows()
        {
            var entity = new TestEntity
            {
                Property = "prop",
                ConcurrencyToken = "19ec62bea4f94ef1a8e4723707d3cc85"
            };

            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                context.Add(entity);
                context.SaveChanges();
            }

            using (var context = InMemoryContextBuilder.Build<TestDataContext>())
            {
                var update = new TestEntity
                {
                    Id = entity.Id,
                    Property = "Something new",
                    ConcurrencyToken = "dbd48fc43d9244f7a08cd54bae94b15a"
                };
                context.Attach(update);
                context.Entry(update).Property("Property").IsModified = true;
                var exception = Assert.Throws<DbUpdateConcurrencyException>(() => context.SaveChanges());
                Approvals.Verify(exception.Message);
            }
        }

        public class TestEntity
        {
            public string ConcurrencyToken { get; set; }
            
            public int Id { get; set; }

            public string Property { get; set; }
        }

        private class TestDataContext : DbContext
        {
            public TestDataContext(DbContextOptions options)
                : base(options)
            { }

            public DbSet<TestEntity> TestEntities { get; set; }

            protected override void OnModelCreating(ModelBuilder modelBuilder)
            {
                var entity = modelBuilder.Entity<TestEntity>();
                entity.HasKey(p => p.Id);

                entity.Property(p => p.Id)
                    .ValueGeneratedOnAdd();

                entity.Property(p => p.ConcurrencyToken)
                    .IsConcurrencyToken();
            }
        }
    }
}