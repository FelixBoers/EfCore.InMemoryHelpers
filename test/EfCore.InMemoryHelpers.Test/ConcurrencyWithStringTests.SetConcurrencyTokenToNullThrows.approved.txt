Attempted to update or delete an entity that does not exist in the store.